### PR TITLE
feat: typed encoder to bridge the gap between EIP-712 and ABI

### PR DESCRIPTION
**Pull request by fish-sammy**
# TypedEncoder: Generic Struct Encoding Library

## Overview

This PR introduces `TypedEncoder`, a new library for generic struct encoding that supports both ABI encoding and EIP-712 structHash computation. The library enables encoding arbitrary struct types at runtime without compile-time type information.

## What is TypedEncoder?

`TypedEncoder` provides a flexible data structure for representing and encoding arbitrary Solidity types dynamically:

```solidity
struct Struct {
    bytes32 typeHash;
    Chunk[] chunks;
}

struct Chunk {
    Primitive[] primitives;
    Struct[] structs;
    Array[] arrays;
}
```

### 📦 ABI Encoding

`abiEncode()` produces standard ABI-encoded output matching Solidity's native `abi.encode()`:
- Static structs: direct encoding without offset wrapper
- Dynamic structs: proper offset wrapper with head/tail separation
- Arrays: correct handling of fixed-size vs dynamic arrays
- Nested structures: recursive encoding with proper offset calculation

### 🔐 EIP-712 StructHash

`structHash()` computes EIP-712 compliant hashes for typed structured data:
- Static types: encoded directly
- Dynamic types: `keccak256()` of data
- Nested structs: recursive structHash computation
- Arrays: `keccak256(abi.encodePacked(...))` of element hashes

### ✅ Supported Types

- Static primitives: `uint256`, `address`, `bytes32`, etc.
- Dynamic primitives: `string`, `bytes`
- Fixed-size arrays: `uint256[3]`, `string[2]`
- Dynamic arrays: `uint256[]`, `string[]`
- Nested structs: arbitrary depth
- Multi-dimensional arrays: `string[][]`
- Mixed static/dynamic fields

#### `TypedEncoderAbiEncode.t.sol` - 19 tests

**Section 1: Basic Primitives** (6 tests)
- Static fields only (`uint256`, `address`)
- Dynamic field only (`string`)
- Mixed static/dynamic
- Fixed bytes (`bytes32`)
- Empty dynamic types
- Complex mixed types

**Section 2: Basic Arrays** (6 tests)
- Static array of static type (`uint256[3]`)
- Static array of dynamic type (`string[2]`)
- Dynamic array of static type (`uint256[]`)
- Dynamic array of dynamic type (`string[]`)
- Empty array
- Single element array

**Section 3: Advanced Arrays** (2 tests)
- Nested arrays (`string[][]`)
- Multiple arrays in struct

**Section 4: Structs** (2 tests)
- Nested struct
- Struct with array field

**Section 5: Arrays of Structs** (2 tests)
- Array of static structs
- Array of dynamic structs

**Section 6: Complex** (1 test)
- Multiple chunks (field ordering)

#### `TypedEncoderStructHash.t.sol` - 9 tests

**Section 1: Basic Primitives** (5 tests)
- Static/dynamic/mixed fields
- Fixed bytes
- Empty dynamic types

**Section 2: Basic Arrays** (1 test)
- Dynamic array of dynamic type

**Section 3: Nested Structs** (1 test)
- Multi-level nesting

**Section 4: Struct with Array** (1 test)
- Array fields within structs

**Section 5: 2D Arrays** (1 test)
- Nested array structures (`string[][]`)

### ✅ Test Results

```
✅ 19/19 ABI encoding tests pass
✅ 9/9 structHash tests pass
✅ 4/4 original TypedEncoder.t.sol tests pass
```

**Testing Methodology**:
- **ABI tests**: Compare TypedEncoder output against Solidity's native `abi.encode()`
- **StructHash tests**: Manually compute expected EIP-712 hash per spec for independent verification

## Files Added

- `src/libs/TypedEncoder.sol` - Core library implementation
- `test/libs/TypedEncoderAbiEncode.t.sol` - ABI encoding test suite
- `test/libs/TypedEncoderStructHash.t.sol` - StructHash test suite

## Use Cases

This library enables:
- **Generic signature verification** without compile-time type knowledge
- **Dynamic ABI encoding** for cross-contract calls
- **Flexible EIP-712 signing** for arbitrary struct types
- **Runtime type composition** for complex data structures
- **Generic permit systems** that work with any token standard

## Example Usage

```solidity
// Create a generic struct representation
TypedEncoder.Struct memory s = TypedEncoder.Struct({
    typeHash: keccak256("Transfer(address to,uint256 amount)"),
    chunks: new TypedEncoder.Chunk[](1)
});
s.chunks[0].primitives = new TypedEncoder.Primitive[](2);
s.chunks[0].primitives[0] = TypedEncoder.Primitive({
    isDynamic: false,
    data: abi.encode(recipient)
});
s.chunks[0].primitives[1] = TypedEncoder.Primitive({
    isDynamic: false,
    data: abi.encode(amount)
});

// Get ABI encoding
bytes memory encoded = s.abiEncode();

// Get EIP-712 hash
bytes32 hash = s.structHash();
```

## Implementation Quality

- ✅ **Correctness**: All tests validate against Solidity's native encoding
- ✅ **Robustness**: Comprehensive edge case coverage
- ✅ **Maintainability**: Clean code organization with clear separation of concerns
- ✅ **Well-tested**: 28 tests covering all scenarios

### CODE REVIEW

File: src/libs/TypedEncoder.sol [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+        Primitive[] primitives;
+        Struct[] structs;
+        Array[] arrays;
+    }
+
+    /// @notice Computes the EIP-712 structHash for signature validation
+    /// @dev Follows EIP-712 specification: keccak256(abi.encodePacked(typeHash, encodeData...))
+    /// @param s The struct to hash
+    /// @return The EIP-712 compliant struct hash
+    function structHash(
```
Comment (re1ro) [inline src/libs/TypedEncoder.sol:48]: rename to `hash()`

```
using TypedEncoder for TypedEncoder.Struct;

struct.hash();
```

**Review by re1ro — APPROVED** [review]
